Sources/GRPCCore/Metadata.swift

```diff
@@ -14,5 +14,267 @@
+    @usableFromInline
+    let _base: Array<Element>.Index
+
+    @inlinable
+    init(_base: Array<Element>.Index) {
+      self._base = _base
+    }
+
+    @inlinable
+    public static func < (lhs: Index, rhs: Index) -> Bool {
+      return lhs._base < rhs._base
+    }
+  }
+
+  public var startIndex: Index {
+    return .init(_base: self.elements.startIndex)
+  }
+
+  public var endIndex: Index {
+    return .init(_base: self.elements.endIndex)
+  }
+
+  public func index(before i: Index) -> Index {
+    return .init(_base: self.elements.index(before: i._base))
+  }
+
+  public func index(after i: Index) -> Index {
+    return .init(_base: self.elements.index(after: i._base))
+  }
+
+  public subscript(position: Index) -> Element {
+    self.elements[position._base]
+  }
+}
+
+extension Metadata {
+
+  /// An iterator for all string values associated with a given key.
+  /// This iterator will only return values originally stored as strings for a given key.
+  public struct StringValuesIterator: IteratorProtocol {
+    private var metadataIterator: Metadata.Iterator
+    private let key: String
+
+    init(forKey key: String, metadata: Metadata) {
+      self.metadataIterator = metadata.makeIterator()
+      self.key = key
+    }
+
+    public mutating func next() -> String? {
+      while let nextKeyValue = self.metadataIterator.next() {
+        if nextKeyValue.key == self.key {
+          switch nextKeyValue.value {
+          case .string(let stringValue):
+            return stringValue
+          case .binary:
+            continue
+          }
+        }
+      }
+      return nil
+    }
+  }
+
+  /// Create a new ``StringValuesIterator`` that iterates over the string values for the given key.
+  /// - Parameter key: The key over whose string values this iterator will iterate.
+  /// - Returns: An iterator to iterate over string values for the given key.
+  public func makeStringValuesIterator(forKey key: String) -> StringValuesIterator {
+    StringValuesIterator(forKey: key, metadata: self)
+  }
+
+  /// A subscript to get a ``StringValuesIterator`` for a given key.
+  public subscript(keyForStringValues key: String) -> StringValuesIterator {
+    StringValuesIterator(forKey: key, metadata: self)
+  }
+}
+
+extension Metadata {
+
+  /// An iterator for all binary data values associated with a given key.
+  /// This iterator will only return values originally stored as binary data for a given key.
+  public struct BinaryValuesIterator: IteratorProtocol {
+    private var metadataIterator: Metadata.Iterator
+    private let key: String
+
+    init(forKey key: String, metadata: Metadata) {
+      self.metadataIterator = metadata.makeIterator()
+      self.key = key
+    }
+
+    public mutating func next() -> Data? {
+      while let nextKeyValue = self.metadataIterator.next() {
+        if nextKeyValue.key == self.key {
+          switch nextKeyValue.value {
+          case .string:
+            continue
```

Comment:
If we receive binary data from the remote peer it'll be base64 encoded, so if we get a string here we should base64 decode it.

Reply:
Hmm shouldn't we do that when adding a new `Metadata.Value` instead, and allowing `Value.binary` to be created from a base64-encoded `String`? Seems a bit weird maybe to do this conversion when iterating? You think a user would be interested in getting a binary/base64 string as a string (using the string values iterator/sequence)?

Reply:
If the transport layer receives binary metadata then they'd have to decode it to binary to add it to the metadata which means that either every transport has to provide base64 decoding or we have to add functionality to `Metadata` which decodes the string when adding it.

The latter is better because we only have to implement it once. However, if we do that then we _always_ pay the cost of decoding _and_ allocating an array, even if the values aren't used. If we do it here we only do it if the users looks for the value.

Reply:
Okay, that's reasonable. I was worried about (a) trying to decode strings that aren't base64 encoded and (b) returning base64 encoded strings as plain strings when iterating over the `stringValues`; but I guess the `-bin` suffix should avoid most of these scenarios. And if it doesn't, it's not really serious anyways.

Reply:
Yup, have no way of knowing beyond "-bin" for values we receive from the remote peer.